### PR TITLE
Improve robustness of Crystal lexer

### PR DESCRIPTION
**Pull request by pyrmont**
Crystal's visual sample was not very comprehensive and meant that the lexer did not cover elements of the syntax (notably macros and `%` conditionals).

This PR adds a more comprehensive visual sample (based on the one in Pygments) and makes changes to the lexer to fix coverage errors.

### CODE REVIEW

**Comment by pyrmont** [comment]
@rymiel I'm not sure if you're in a position to do so but if you could have a peek at this, it'd be much appreciated. I've added a more comprehensive visual sample for the Crystal lexer and tried to fix some of the areas in which the lexer was deficient.

File: spec/visual/samples/crystal [hunk 1]
```diff
@@ -1,45 +1,687 @@
...
-    ret
+  class CurrencyConversion
+    def initialize(@amount, @target_currency)
+    end
+
+    def amount
+      # implement conversion ...
+    end
   end
+Vend
```
Comment (rymiel) [inline spec/visual/samples/crystal:102]: Is this intentional? Seems like it'd create a module that is never `end`ed but I am unsure of the spec goal

**Review by rymiel — COMMENTED** [review]
I am in no way in any authoritative position on this, I simply was annoyed that half of my files were turning bland, though, I read through it from a simple user's perspective, I can't immediately think of any language constructs missing, though I'll hopefully continue to participate on this topic.

File: spec/visual/samples/crystal (hunk 1 above)
Comment (pyrmont) [reply spec/visual/samples/crystal]: Oops! Nope, didn't mean to have that there. Nice catch!

**Comment by pyrmont** [comment]
@rymiel Cool. Well, I think I'll merge it in sometime tomorrow and if there are any issues, I'm sure they'll be reported soon enough. Thanks for taking a look!

**Comment by rymiel** [comment]
I tried *really* hard to break it but the lexer is very robust as of now. Syntax that breaks other lexers like the one used in github doesn't break rouge, however, after trying really hard and going back to what broke it before, *regex literals*, I've found that **nested regex interpolation** `/#{/#{nested}/}/` does not work correctly. Yes, it's ridiculous, but [it compiles](https://play.crystal-lang.org/#/r/a2sg). Note that whatever the crystal playground uses [is less robust than rouge](https://user-images.githubusercontent.com/53803019/101384325-f333c800-38c2-11eb-9eb5-936849c7b409.png)

I haven't investigated the cause yet but mostly I just wanted to say how much I struggled to find any problems.

**Comment by pyrmont** [comment]
@rymiel That's great to hear. I can't really take any credit, the Crystal lexer is very closely based on Rouge's Ruby lexer and I think it's the best around (probably not a surprise for a Ruby lexing library). I'll probably leave the nested regular expressions for now and see if it comes up naturally as an issue.

Thanks for giving it such a thorough test! It was really helpful :)

**Pull request merged**